src/amltk/pytorch/builders.py

```diff
@@ -0,0 +1,167 @@
+            choices (dict): A dictionary containing dimensions for choices.
+        """
+        self.choice_name = choice_name
+        self.choices = choices
+
+    def evaluate(self, chosen_nodes) -> int:
+        """Retrieves the corresponding dimension for the chosen node.
+
+        If the chosen node is not found in the choices dictionary, an error is raised.
+        If the dimensions provided are not valid, an error is not raised.
+        It is up to the user to ensure that the dimensions are valid.
+
+        Args:
+            chosen_nodes: The chosen nodes.
+
+        Returns:
+            The value of the matching dimension for a chosen node.
+        """
+        chosen_node_name = chosen_nodes.get(self.choice_name)
+
+        try:
+            return self.choices[chosen_node_name]
+        except KeyError:
+            raise MatchChosenDimensionsError(self.choice_name, chosen_node_name)
+
+    @staticmethod
+    def collect_chosen_nodes_names(pipeline: Sequential) -> dict:
+        """Collects the names of chosen nodes in the pipeline.
+
+        Each pipeline has a unique set of chosen nodes, which we collect separately
+        to handle dimension matching between layers with search spaces.
+
+        Args:
+            pipeline: The pipeline containing the model architecture.
+
+        Returns:
+            The names of the chosen nodes in the pipeline.
+        """
+        chosen_nodes_names = {}  # Class variable to store chosen node names
+
+        for node in pipeline.iter():
+            if isinstance(node, Choice):
+                chosen_node = node.chosen()
+                if chosen_node:
+                    chosen_nodes_names[node.name] = chosen_node.name
+
+        return chosen_nodes_names
+
+
+def build_model_from_pipeline(pipeline: Sequential) -> nn.Module:
+    """Builds a model from the provided pipeline.
+
+    This function iterates through the pipeline nodes, constructing the model
+    layers dynamically based on the node types and configurations. It also
+    utilizes the `MatchDimensions` and `MatchChosenDimensions` objects to handle
+    dimension matching between layers with search spaces.
+
+    Args:
+        pipeline: The pipeline containing the model architecture.
+
+    Returns:
+        The constructed PyTorch model.
+    """
+    model_layers = []
+
+    # Collect the names of chosen nodes for the given pipeline
+    chosen_nodes_names = MatchChosenDimensions.collect_chosen_nodes_names(pipeline)
+
+    for node in pipeline.iter(skip_unchosen=True):
+        # Check if node is a Fixed layer (e.g., Flatten, ReLU)
+        if isinstance(node, Fixed):
+            model_layers.append(node.item)
+
+        # Check if node is a Component with config parameter
+        elif (
+            isinstance(node.item, type)
+            and issubclass(node.item, nn.Module)
+            and node.config
+        ):
+            layer_config = node.config or {}
```

Comment:
So there's a silent potential bug here in that your assign `layer_config` to point to `node.config` and then update `layer_config` with new keys, essentially inserting new keys into `node.config`. I doubt this is likely a bug that would come up but it goes against the more _immutable_ nature of the pipelines. 

Main thing would be to just copy the `node.config` which you can easily do with a `Mapping/dict` type by just calling `dict` on it.

```python
        elif (
            isinstance(node.item, type)
            and issubclass(node.item, nn.Module)
            and node.config is not None
        ):
        	layer_config = dict(node.config)  # Does a shallow copy so you can update the keys
```